### PR TITLE
lkl: use language-specific TLS implementation

### PR DESCRIPTION
**Pull request by xjia1**
Use language-specific TLS implementation instead of the Windows or
pthreads implementation.

The initial motivation for this change is because glibc dlmopen()
cannot properly initialize pthread TLS to NULL.  When combined with
auto_syscall_threads=1 this leads to memory access errors because
the code in arch/lkl/kernel/syscalls.c would assume the per-thread
syscall thread is already created, while the TLS data is actually
a random number.

glibc bug is https://sourceware.org/bugzilla/show_bug.cgi?id=15134

pthreads TLS support is also not available on some platforms, even
without the use of dlmopen().

Replacing the platform-provided TLS stuff with __thread gives us
the extra benefit that the host dependency is reduced.

Signed-off-by: Xiao Jia xiaoj@google.com

---

This change is [Reviewable](https://reviewable.io/reviews/lkl/linux/138)

### CODE REVIEW

**Comment by pscollins** [comment]
LGTM. I think this is much cleaner and more portable.

Personally I'd like it if there wasn't any distinction between "default" and non-default syscall thread because the default is just the syscall thread that happens to serve requests for the first thread. That's a slightly involved change to make though, so it can wait for another time.

File: arch/lkl/kernel/syscalls.c [hunk 1]
```diff
@@ -342,24 +336,17 @@ int initial_syscall_thread(void *sem)
...
-	if (ret)
-		return ret;
-
 	init_pid_ns.child_reaper = 0;
 
 	ret = syscall_thread_data_init(&default_syscall_thread_data, sem);
 	if (ret)
 		goto out;
 
+	syscall_thread_data = &default_syscall_thread_data;
```
Comment (liuyuan10) [inline arch/lkl/kernel/syscalls.c]: I think this is unnecessary. default syscall thread is a special kernel thread that is not attached to any host thread. Also here, it's a kernel thead instead of host thread running this function.

**Comment by liuyuan10** [comment]
@pscollins The default sys thread is actually special. It's created by kernel init thread instead of a host thread and solely works on creating new sys threads and stop them. I won't say it's necessary to have it, but there are reasons to. For example, when a host thread wants to create a new syscall thread, which existing sys thread can serve it? Having a default one to handle it is an easy solution.

**Comment by thehajime** [comment]
I like this cleanup.

I guess it's much cleaner (so I agree with @pscollins) but less portable, OTOH.

I haven't deeply investigated yet but  ```__thread``` storage class implementation for custom thread implementation (like I'm using with ```ucontext``` based one) involves build toolchains while previous implementation using host_ops member to access TLS is trivial.

The custom thread implementation is not upstreamed yet so, I don't disagree anymore here but just wanted to comment this. It would be nice to hear if you have nice solution for such an immature thread implementation.

so basically I'm on +1 for this PR.

**Comment by pscollins** [comment]
@thehajime `__thread` can compile down to `pthread_tls_get` on some platforms, but also to the Win32 equivalent, or a more optimized version on platforms that support it.

I'm not too positive about this, but I'm not sure if ucontext-based parallelism would have any need for multiple syscalls threads

File: arch/lkl/kernel/syscalls.c (hunk 1 above)
Comment (stfairy) [inline arch/lkl/kernel/syscalls.c]: Done.

**Comment by stfairy** [comment]
@thehajime Thanks for the comment.  As I mentioned in the commit message, this is not only a cleanup, but also a bug fix/workaround for the dlmopen+TLS implementation in glibc.  We could figure out how to do all of this more portable in a future commit I guess.

**Comment by stfairy** [comment]
Actually, I'd rather we hold on this commit as it introduces a new problem: I'm getting the error "cannot load any more object with static TLS".

**Comment by stfairy** [comment]
Well, the error seems to come from dlmopen(librt.so) when it's opened the second time.  In that case it seems not feasible to load two instances of LKL in a same process because of the limited static TLS space.

By examining source code it seems that this error message is removed from glibc-2.22, but I doubt if it's really resolved.

**Comment by pscollins** [comment]
@stfairy I think that if we go with the original API `tls_set()`, `tls_get()`, etc we could pretty easily roll our own TLS as an array of `struct syscall_thread_data` keyed on the index of the syscall thread. I can throw it together later if you think it's worth exploring.

**Comment by stfairy** [comment]
@pscollins That doesn't help with the librt which we can't really change.

I'll close this PR for now.
